csharp/ql/src/semmle/code/csharp/dataflow/FrameworkDataFlow.qll

```diff
@@ -0,0 +1,502 @@
+/**
+ * Provides classes and predicates for definining data-flow through frameworks.
+ */
+
+import csharp
+private import semmle.code.csharp.dataflow.internal.DataFlowPrivate
+private import semmle.code.csharp.dataflow.internal.DataFlowPublic
+private import semmle.code.csharp.frameworks.system.Collections
+private import semmle.code.csharp.frameworks.system.collections.Generic
+private import semmle.code.csharp.frameworks.system.linq.Expressions
+
+/** An unbound callable. */
+class SourceDeclarationCallable extends Callable {
+  SourceDeclarationCallable() { this = this.getSourceDeclaration() }
+}
+
+/** An unbound method. */
+class SourceDeclarationMethod extends SourceDeclarationCallable, Method { }
+
+private predicate hasDelegateArgumentPosition(SourceDeclarationCallable c, int i) {
+  exists(DelegateType dt |
+    dt = c.getParameter(i).getType().(SystemLinqExpressions::DelegateExtType).getDelegateType()
+  |
+    not dt.getReturnType() instanceof VoidType
+  )
+}
+
+private predicate hasDelegateArgumentPosition2(SourceDeclarationCallable c, int i, int j) {
+  exists(DelegateType dt |
+    dt = c.getParameter(i).getType().(SystemLinqExpressions::DelegateExtType).getDelegateType()
+  |
+    exists(dt.getParameter(j))
+  )
+}
+
+/** INTERNAL: Do not use. */
+module Internal {
+  newtype TContentStack =
```

Comment:
What distinguishes a `TContentStack` from a `DataFlow::TAccessPath`?

Reply:
`AccessPath`s are configuration-dependant, and they also annotate each content with a type.